### PR TITLE
[Docs only] Better docstrings for task, workflow, dynamic

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Docs only.

### CODE REVIEW

File: flytekit/core/dynamic_workflow_task.py [hunk 1]
```diff
@@ -40,3 +40,35 @@ def __init__(
 dynamic = functools.partial(
     task.task, task_config=_Dynamic(), execution_mode=PythonFunctionTask.ExecutionBehavior.DYNAMIC
 )
+dynamic.__doc__ = """
+Please first see the comments for :py:func:`flytekit.task` and :py:func:`flytekit.workflow`. This ``dynamic``
+concept is an amalgamation of both and enables the user to pursue some :std:ref:`pretty incredible <cookbook:advanced_merge_sort>`
```
Comment (katrogan) [inline flytekit/core/dynamic_workflow_task.py:45]: 😄

File: flytekit/core/task.py [hunk 2]
```diff
@@ -69,58 +69,62 @@ def task(
     **kwargs,
 ) -> Union[Callable, PythonFunctionTask]:
     """
-    This is the core decorator to use for any task type in FlyteKit.
-    Usage: for a simple python task
+    This is the core decorator to use for any task type in flytekit. For a simple python task,
```
Comment (katrogan) [inline flytekit/core/task.py:72]: did you mean to add more at the end here?

File: flytekit/core/task.py [hunk 3]
```diff
@@ -69,58 +69,62 @@ def task(
...
+    Please see some cookbook :std:ref:`task examples <cookbook:tasks>` for additional information.
 
     :param _task_function: This argument is implicitly passed and represents the decorated function
     :param task_config: This argument provides configuration for a specific task types.
-                        Please refer to the plugins documentation for the right Object to use
+                        Please refer to the plugins documentation for the right object to use.
     :param cache: Boolean that indicates if caching should be enabled
-    :param cache_version: Version string to be used for the cached value
-    :param retries: for retries=n; n > 0, on failures of this task, the task will be retried at-least n number of times.
+    :param cache_version: Cache version to use. Changes to the task signature will automatically trigger a cache miss,
```
Comment (katrogan) [inline flytekit/core/task.py:94]: Can we add that you *should* bump it for changed task bodies?

**Review by katrogan — DISMISSED** [review]

File: flytekit/core/task.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/core/task.py:72]: no, usage follows, but I'm using a code block, so I can't use ::

File: flytekit/core/task.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/core/task.py:94]: yes done

File: flytekit/core/dynamic_workflow_task.py [hunk 4]
```diff
@@ -40,3 +40,35 @@ def __init__(
...
+constructs.
+
+In short, a task's function is run at execution time only, and a workflow function is run at compilation time only (local
+execution notwithstanding). A dynamic workflow is modeled on the backend as a task, but at execution time, the function
+body gets compiled like a workflow, as if the decorator changed from ``@task`` to ``@workflow``. This workflow is
```
Comment (kumare3) [inline flytekit/core/dynamic_workflow_task.py:50]: This is not accurate- if you change task to workflow you cannot use inputs, while in dynamic you can

**Review by kumare3 — APPROVED** [review]

File: flytekit/core/dynamic_workflow_task.py (hunk 4 above)
Comment (wild-endeavor) [reply flytekit/core/dynamic_workflow_task.py:50]: you mean you can't use inputs as if they're native python things?  i'll clarify

**Review by katrogan — APPROVED** [review]

**Pull request merged**